### ISSUE TITLE
Black image when capturing an iframe of Google Street View

### ISSUE THREAD

**Issue by AnnoyingTechnology** [1904labs/dom-to-image-more#50]
### Use case: description, code  

Thanks for the help.

```js
domtoimage.toJpeg(
		document.getElementById('street_view'), 
		{ 'quality':0.95 }
	).then(function (dataUrl) {
		
		var img = new Image();
		img.src = dataUrl;
		document.body.appendChild(img);
	})
	.catch(function (error) {
		console.error("oops, something went wrong!", error);
	});
```

```html
<iframe 
id="street_view" 
width="100%"
height="1024px"
style="border:0"
loading="lazy"
allowfullscreen
src="https://www.google.com/maps/embed/v1/streetview?key=REDACTED
&location=47.2740517,-2.2116541
&heading=10
&pitch=10
&fov=70">
</iframe>
```

### Expected behavior

An image containing the screenshot of the iframe

### Actual behavior (stack traces, console logs etc)

Image  of the correct size, but entirely black.
No exception are thrown, no errors.

The only error is `js?client=google-maps-embed&paint_origin=&libraries=geometry,search&v=3.exp&language=en_US&callback=onApiLoad:258 The deviceorientation events are blocked by permissions policy.`

### Library version

master & dom-to-image-more 21-05-2019 

### Browsers

- [x] Chrome 49+  (`90.0.4430.212 `)

**Comment by southpush** [comment]
Try to use toPng, I get a black picture too when I use toJpeg, but work well in toPng

**Comment by AnnoyingTechnology** [comment]
Thanks for the tip. But this had already been tried without success.

**Comment by ransomsumit** [comment]
Were you able to solve this issue @AnnoyingTechnology ?

**Comment by AnnoyingTechnology** [comment]
> Were you able to solve this issue @AnnoyingTechnology ?

Hi, sadly no. I had to use a workaround without using an iframe.

**Comment by IDisposable** [comment]
Does adding ` allow="accelerometer; magnetometer; gyroscope"` to the iframe work? That would allow the device sensors inside third-party domain loaded scripts.

**Comment by ransomsumit** [comment]
Hey Guys, i just found a big bug with the instanceof checks in the source code. 
Trying instanceof with a dom element situated inside an iframe on window prototypes  gives false on chrome but gives true on firefox. 

Fixing this solved my issue with iframes.

**Comment by IDisposable** [comment]
@ransomsumit Did you have a PR or a general fix or more details you would like to contribute?

**Comment by ransomsumit** [comment]
@IDisposable Here is the fix 
https://github.com/ransomsumit/dom-to-image/blob/master/src/dom-to-image.js

**Comment by IDisposable** [comment]
Thanks! I've generalized it a bit more to reach up into the contentWindow as needed. Can you confirm that the version in PR #80 works for you so I can make a general release?

**Comment by ransomsumit** [comment]
No it doesn't work properly. it's missing several styles. @IDisposable

**Comment by IDisposable** [comment]
Any suggestion as to what we're missing? I would love to roll out a fix.

**Comment by ransomsumit** [comment]
The one thing I noticed is that some tags For ex. <g> gives out false on instanceof check within the frame for some reason but gives true on root window. So the fail safe will be to check instanceof on root window and the frame window at the same time with proper condition. My fix is working fine with this. 

Currently,
I am tackling some weird image event error now, if you have any context please tell. @IDisposable

**Comment by IDisposable** [comment]
I've pushed up some more extensive cleanup, mostly around the finding of the owner window object and also fixed a really silly issue with class name generation used when copying over any `::before` or `::after` attributes. 

Please pull the branch again and see if I've got it working. If you can give me a repro that show the error, I can fix it faster (follow the pattern in the unit testes... an HTML, a CSS, (optionally) a control-image.

**Comment by zoltanseer** [comment]
Maybe it's because iframe has `loading=lazy` attribute

**Comment by meche-gh** [comment]
Parallel to the `iframe` feature request:

\`getComputedStyle\` recomputes layout. We want to target recomputes to the element's window, not \`global || window\`. I believe the internal implementation in Blink:

*   reaches up from \`element.ownerDocument.defaultView\` back to `global || window` making this change a 5-10% slight slowdown 
*   also that `window.getComputedStyle` triggers a layout recompute in iframes anyway

It just seems more robust to be using the window that's the immediate context of the element itself. We could do this now or when we implement iframes.. if the latter, I'll log a new issue.

**Comment by meche-gh** [comment]
For iframe cloning:

*   We should first check if there is an srcdoc because that is usually inlined as a higher priority than the `<iframe>` `src`.
*   Same origin = `iframe.contentDocument` is available: we can attempt a jump from `iframe` to `iframe.contentDocument` when getting the children. Similar to #101.
*   Otherwise there should be the option to fetch the HTML of `src` via XHR and inline it.

**Comment by meche-gh** [comment]
For iframe cloning:

*   We should first check if there is an srcdoc because that is usually inlined as a higher priority than the `<iframe>` `src`.
*   Same origin = `iframe.contentDocument` is available: we can attempt a jump from `iframe` to `iframe.contentDocument` when getting the children. Similar to #101.
*   Otherwise there should be the option to fetch the HTML of `src` via XHR and inline it.

**Comment by meche-gh** [comment]
Resources inside the cloned iframe document don't load after the XHR. So cross-origin support is a no-no.
I believe Google Street View is always a cross-origin embed/iframe so it's impossible to take a screenshot of without retooling some kind of [advanced browser API](https://developer.mozilla.org/en-US/docs/Web/API/Screen_Capture_API/Using_Screen_Capture) and requiring user input after the `toSvg` call.

**Comment by IDisposable** [comment]
Street View renders in a cross-origin `<iframe>` (and via WebGL), and the browser security model makes both uncapturable from the host page — the library can't reach inside a _cross-origin_ `<iframe>` or read an unpreserved GL buffer, so it comes out black. This isn't fixable from our side. If you control the embed, render a static Street View image (the Static API) into a **same-origin** element and capture that instead. Closing as a known limitation — thanks for the detailed thread.

**Comment by IDisposable** [comment]
Street View renders in a cross-origin `<iframe>` (and via WebGL), and the browser security model makes both uncapturable from the host page — the library can't reach inside a _cross-origin_ `<iframe>` or read an unpreserved GL buffer, so it comes out black. This isn't fixable from our side. If you control the embed, render a static Street View image (the Static API) into a **same-origin** element and capture that instead. Closing as a known limitation — thanks for the detailed thread.
